swarm/src/protocols_handler/either.rs

```diff
@@ -0,0 +1,329 @@
+// Copyright 2021 Protocol Labs.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use crate::protocols_handler::{
+    IntoProtocolsHandler, KeepAlive, ProtocolsHandler, ProtocolsHandlerEvent,
+    ProtocolsHandlerUpgrErr, SubstreamProtocol,
+};
+use crate::upgrade::{InboundUpgradeSend, OutboundUpgradeSend, SendWrapper};
+use either::Either;
+use libp2p_core::either::{EitherError, EitherOutput};
+use libp2p_core::upgrade::{EitherUpgrade, UpgradeError};
+use libp2p_core::{ConnectedPoint, Multiaddr, PeerId};
+use std::task::{Context, Poll};
+
+pub enum IntoEitherHandler<A, B> {
+    A(A),
+    B(B),
+}
+
+/// Implementation of a [`IntoProtocolsHandler`] that represents either of two [`IntoProtocolsHandler`]
+/// implementations.
+impl<A, B> IntoProtocolsHandler for IntoEitherHandler<A, B>
+where
+    A: IntoProtocolsHandler,
+    B: IntoProtocolsHandler,
+{
+    type Handler = EitherHandler<A::Handler, B::Handler>;
+
+    fn into_handler(self, p: &PeerId, c: &ConnectedPoint) -> Self::Handler {
+        match self {
+            IntoEitherHandler::A(into_handler) => EitherHandler::A(into_handler.into_handler(p, c)),
+            IntoEitherHandler::B(into_handler) => EitherHandler::B(into_handler.into_handler(p, c)),
+        }
+    }
+
+    fn inbound_protocol(&self) -> <Self::Handler as ProtocolsHandler>::InboundProtocol {
+        match self {
+            IntoEitherHandler::A(into_handler) => {
+                EitherUpgrade::A(SendWrapper(into_handler.inbound_protocol()))
+            }
+            IntoEitherHandler::B(into_handler) => {
+                EitherUpgrade::B(SendWrapper(into_handler.inbound_protocol()))
+            }
+        }
+    }
+}
+
+/// Implementation of a [`ProtocolsHandler`] that represents either of two [`ProtocolsHandler`]
+/// implementations.
+#[derive(Debug)]
+pub enum EitherHandler<A, B> {
+    A(A),
+    B(B),
+}
+
+impl<A, B> ProtocolsHandler for EitherHandler<A, B>
```

Comment:
If _either_ (pun intended) https://github.com/bluss/either/issues/57 gets resolved or we copy-paste the macro, a lot of these implementations would get simpler if they were all based on the `Either` type instead of custom ones.